src/dotty/tools/dotc/reporting/diagnostic/messages.scala

```diff
@@ -512,4 +513,64 @@ object messages {
         |}""".stripMargin
   }
 
+  case class WrongNumberOfArgs(fntpe: Type, argKind: String, expectedArgs: List[TypeParamInfo], actual: List[untpd.Tree])(implicit ctx: Context)
+    extends Message(18) {
+    val kind = "Syntax"
+    val expectedCount = expectedArgs.length
+    val actualCount = actual.length
+    val msgPrefix = if (actualCount > expectedCount) "Too many" else "Not enough"
+
+    val expectedArgString = (fntpe.widen match {
+      case pt: MethodOrPoly => pt //ensure we return a type that will have useful typeParms
+      case _ => fntpe
+    }).typeParams.map(_.paramName.show.split("\\$").last).mkString("[", ", ", "]")
+
+    val actualArgString = actual.map(_.show.split("\\.").last).mkString("[", ", ", "]")
```

Comment:
I remember why I had no luck with asSeenFrom, one of the four callers of this function has an untyped arg tree.  Should I split the message into two, one for typed callers and one for untyped callers, or just leave the full path?